src/accel/detail/IntegrationSingleton.cc

```diff
@@ -0,0 +1,231 @@
+    };
+
+    CELER_TRY_HANDLE(initialize_impl(),
+                     ExceptionConverter{"celer.init.logger"});
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct shared params on master (or single) thread.
+ *
+ * \todo The query for CeleritasDisabled initializes the environment before
+ * we've had a chance to load the user setup options. Make sure we can update
+ * the environment *first* when refactoring the setup.
+ *
+ * \note In Geant4 threading, \em only MT mode on non-master thread has
+ *   \c G4Threading::IsWorkerThread()==true. For MT mode, the master thread
+ *   does not track any particles. For single-thread mode, the master thread
+ *   \em does do work.
+ */
+void IntegrationSingleton::initialize_shared_params()
+{
+    ExceptionConverter call_g4exception{"celer.init.global"};
+
+    if (!G4Threading::IsMultithreadedApplication()
+        || G4Threading::IsMasterThread())
+    {
+        CELER_LOG_LOCAL(debug) << "Initializing shared params";
+
+        auto initialize_impl = [this] {
+            CELER_VALIDATE(
+                !params_,
+                << R"(BeginOfRunAction cannot be called more than once)");
+            params_.Initialize(options_);
+        };
+
+        CELER_TRY_HANDLE(initialize_impl(), call_g4exception);
+    }
+    else
+    {
+        CELER_LOG_LOCAL(debug) << "Initializing worker";
+
+        auto initialize_impl = [this] {
+            CELER_VALIDATE(
+                params_,
+                << R"(BeginOfRunAction was not called on master thread)");
+            params_.InitializeWorker(options_);
+        };
+
+        CELER_TRY_HANDLE(initialize_impl(), call_g4exception);
+    }
+
+    CELER_ENSURE(params_);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct thread-local transporter.
+ *
+ * Note that this uses the thread-local static data. It *must not* be called
+ * from the master thread in a multithreaded run.
+ *
+ * \return Whether Celeritas offloading is enabled
+ */
+bool IntegrationSingleton::initialize_local_transporter()
+{
+    CELER_EXPECT(params_);
+
+    if (G4Threading::IsMultithreadedApplication()
+        && G4Threading::IsMasterThread())
+    {
+        // Cannot construct local transporter on master MT thread
+        return false;
+    }
+
+    CELER_EXPECT(!G4Threading::IsMultithreadedApplication()
+                 || G4Threading::IsWorkerThread());
+
+    if (params_.mode() == celeritas::SharedParams::Mode::disabled)
+    {
+        CELER_LOG_LOCAL(debug)
+            << R"(Skipping state construction since Celeritas is completely disabled)";
+        return false;
+    }
+
+    if (params_.mode() == celeritas::SharedParams::Mode::kill_offload)
+    {
+        // When "kill offload", we still need to intercept tracks
+        CELER_LOG_LOCAL(debug)
+            << R"(Skipping state construction with offload enabled)";
```

Comment:
Maybe make it clear in the log message that nothing is being offloaded in this case?